include/libint2/dfbs_generator.h

```diff
@@ -0,0 +1,305 @@
+/*
+ *  Copyright (C) 2004-2021 Edward F. Valeev
+ *
+ *  This file is part of Libint.
+ *
+ *  Libint is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Lesser General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Libint is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with Libint.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef _libint2_src_lib_libint_dfbs_generator_h_
+#define _libint2_src_lib_libint_dfbs_generator_h_
+
+#include <libint2.h>
+#include <libint2/atom.h>
+#include <libint2/basis.h>
+#include <libint2/boys.h>
+#include <libint2/pivoted_cholesky.h>
+#include <math.h>
+
+#include <Eigen/Dense>
+#include <Eigen/Eigenvalues>
+#include <algorithm>
+
+namespace libint2 {
+
+namespace detail {
+
+/// Computes uncontracted shells from a vector of shells
+/// @param[in] cluster a vector of shells
+/// @return a vector of uncontracted shells
+inline std::vector<Shell> uncontract(const std::vector<Shell> &shells) {
+  std::vector<Shell> primitive_shells;
+  for (const auto &contracted_shell : shells) {
+    for (size_t p = 0; p < contracted_shell.nprim(); p++) {
+      const auto prim_shell = contracted_shell.extract_primitive(p, true);
+      // if dealing with generally contracted basis (e.g., cc-pvxz) represented
+      // as a segmented basis need to remove duplicates
+      if (std::find(primitive_shells.begin(), primitive_shells.end(),
+                    prim_shell) == primitive_shells.end())
+        primitive_shells.emplace_back(std::move(prim_shell));
+    }
+  }
+  return primitive_shells;
+}
+
+/// @brief returns \Gamma(x)  of x
+inline double gamma_function(const double x) { return std::tgamma(x); }
+
+/// @brief return effective exponent of product of two primitive shells
+/// @param shell1 first shell
+/// @param shell2 second shell
+/// @param L total angular momentum of product function
+/// @return effective exponent of product function
+inline double alpha_eff(const Shell &shell1, const Shell &shell2, const int L) {
+  auto alpha1 = shell1.alpha[0];
```

Comment:
for the future: most of the time when you write `auto x = ...` (and this is almost always a good idea; see https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/) almost always you want `const auto x = ...` instead. So train yourself to write const auto instinctively.